service/src/test/java/bio/terra/workspace/service/workspace/flight/RemoveUserFromWorkspaceFlightTest.java

```diff
@@ -0,0 +1,188 @@
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledIfEnvironmentVariable;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class RemoveUserFromWorkspaceFlightTest extends BaseConnectedTest {
+
+  @Autowired private WorkspaceService workspaceService;
+  @Autowired private ControlledResourceService controlledResourceService;
+  @Autowired private JobService jobService;
+  @Autowired private SamService samService;
+  @Autowired private SpendConnectedTestUtils spendUtils;
+  @Autowired private UserAccessUtils userAccessUtils;
+
+  private static final Duration STAIRWAY_FLIGHT_TIMEOUT = Duration.ofMinutes(3);
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = BUFFER_SERVICE_DISABLED_ENVS_REG_EX)
+  void removeUserFromWorkspaceFlightDoUndo() throws Exception {
+    // Create a workspace as the default test user
+    WorkspaceRequest request =
+        WorkspaceRequest.builder()
+            .workspaceId(UUID.randomUUID())
+            .workspaceStage(WorkspaceStage.MC_WORKSPACE)
+            .spendProfileId(Optional.of(spendUtils.defaultSpendId()))
+            .build();
+    UUID workspaceId =
+        workspaceService.createWorkspace(request, userAccessUtils.defaultUserAuthRequest());
+    // Add the secondary test user as a writer
+    samService.grantWorkspaceRole(
+        workspaceId,
+        userAccessUtils.defaultUserAuthRequest(),
+        WsmIamRole.WRITER,
+        userAccessUtils.getSecondUserEmail());
+
+    // Create a GCP context as default user
+    String makeContextJobId = RandomStringUtils.randomAlphabetic(8);
+    workspaceService.createGcpCloudContext(
+        workspaceId, makeContextJobId, userAccessUtils.defaultUserAuthRequest());
+    jobService.waitForJob(makeContextJobId);
+    AsyncJobResult<GcpCloudContext> createContextJobResult =
+        jobService.retrieveAsyncJobResult(
+            makeContextJobId, GcpCloudContext.class, userAccessUtils.defaultUserAuthRequest());
+    assertEquals(StatusEnum.SUCCEEDED, createContextJobResult.getJobReport().getStatus());
+    GcpCloudContext gcpContext = createContextJobResult.getResult();
+
+    // Create a private dataset for secondary user
+    String datasetId = RandomStringUtils.randomAlphabetic(8);
+    ControlledBigQueryDatasetResource privateDataset = buildPrivateDataset(workspaceId, datasetId);
+    assertNotNull(privateDataset);
+
+    // Validate with Sam that secondary user can read their private resource
+    assertTrue(
+        samService.isAuthorized(
+            userAccessUtils.secondUserAccessToken().getTokenValue(),
+            privateDataset.getCategory().getSamResourceName(),
+            privateDataset.getResourceId().toString(),
+            SamControlledResourceActions.WRITE_ACTION));
```

Comment:
Should this be `READ_ACTION`?

Reply:
Either is ok. The user was added to the workspace as a writer, so they should still have both permissions at this point.